src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperScopeManager.cs

```diff
@@ -115,14 +114,61 @@ public TagHelperExecutionContext End()
                         nameof(TagHelperScopeManager)));
             }
 
-            _executionScopes.Pop();
+            _executionContextPool.ReturnCurrent();
+
+            var parentExecutionContext = _executionContextPool.Current;
+
+            return parentExecutionContext;
+        }
+
+        private class ExecutionContextPool
+        {
+            private readonly List<TagHelperExecutionContext> _executionContexts;
+            private int _nextIndex;
+
+            public ExecutionContextPool()
+            {
+                _executionContexts = new List<TagHelperExecutionContext>();
+            }
+
+            public TagHelperExecutionContext Current => _nextIndex > 0 ? _executionContexts[_nextIndex - 1] : null;
 
-            if (_executionScopes.Count != 0)
+            public TagHelperExecutionContext Rent(
+                string tagName,
+                TagMode tagMode,
+                IDictionary<object, object> items,
+                string uniqueId,
+                Func<Task> executeChildContentAsync,
+                Action<HtmlEncoder> startTagHelperWritingScope,
+                Func<TagHelperContent> endTagHelperWritingScope)
             {
-                return _executionScopes.Peek();
+                TagHelperExecutionContext tagHelperExecutionContext;
+
+                if (_nextIndex == _executionContexts.Count)
+                {
+                    tagHelperExecutionContext = new TagHelperExecutionContext(
+                        tagName,
+                        tagMode,
+                        items,
+                        uniqueId,
+                        executeChildContentAsync,
+                        startTagHelperWritingScope,
+                        endTagHelperWritingScope);
+
+                    _executionContexts.Add(tagHelperExecutionContext);
+                }
+                else
+                {
+                    tagHelperExecutionContext = _executionContexts[_nextIndex];
```

Comment:
Since `TagHelperScopeManager.End` does the throw (with a test that verifies it throws) I think it'd be a bit much to also have a `Debug.Assert` here